components/raftstore/src/errors.rs

```diff
@@ -30,123 +30,91 @@ pub enum DiscardReason {
     Full,
 }
 
-quick_error! {
-    #[derive(Debug)]
-    pub enum Error {
-        ProposalInMergingMode(region_id: u64) {
-            display("{} peer in merging mode, can't do proposal", region_id)
-        }
-        ReadIndexNotReady(reason: &'static str, region_id: u64) {
-            display("read index not ready, reason {}, region {}", reason, region_id)
-        }
-        RaftEntryTooLarge(region_id: u64, entry_size: u64) {
-            display("raft entry is too large, region {}, entry size {}", region_id, entry_size)
-        }
-        StoreNotMatch(to_store_id: u64, my_store_id: u64) {
-            display("to store id {}, mine {}", to_store_id, my_store_id)
-        }
-        RegionNotFound(region_id: u64) {
-            display("region {} not found", region_id)
-        }
-        RegionNotInitialized(region_id: u64) {
-            display("region {} not initialized yet", region_id)
-        }
-        NotLeader(region_id: u64, leader: Option<metapb::Peer>) {
-            display("peer is not leader for region {}, leader may {:?}", region_id, leader)
-        }
-        KeyNotInRegion(key: Vec<u8>, region: metapb::Region) {
-            display("key {} is not in region key range [{}, {}) for region {}",
-                log_wrappers::Value::key(key),
-                log_wrappers::Value::key(region.get_start_key()),
-                log_wrappers::Value::key(region.get_end_key()),
-                region.get_id())
-        }
-        DataIsNotReady(region_id: u64, peer_id: u64, safe_ts: u64) {
-            display("peer {} is not ready, max_ts {}, region {}", peer_id, safe_ts, region_id)
-        }
-        Other(err: Box<dyn error::Error + Sync + Send>) {
-            from()
-            cause(err.as_ref())
-            display("{:?}", err)
-        }
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("{0} peer in merging mod, can't do proposal")]
+    ProposalInMergingMode(u64),
+    #[error("read index not ready, reason {0}, region {1}")]
+    ReadIndexNotReady(&'static str, u64),
+    #[error("raft entry is too large, region {0}, entry size {1}")]
+    RaftEntryTooLarge(u64, u64),
+    #[error("to store id {0}, mine {1}")]
+    StoreNotMatch(u64, u64),
+    #[error("region {0} not found")]
+    RegionNotFound(u64),
+    #[error("region {0} not initialized yet")]
+    RegionNotInitialized(u64),
+    #[error("peer is not leader for region {0}, leader may {1:?}")]
+    NotLeader(u64, Option<metapb::Peer>),
+    #[error(
+        "key {} is not in region key range [{}, {}), for region {}", 
+        log_wrappers::Value::key(.0),
+        log_wrappers::Value::key(.1.get_start_key()),
+        log_wrappers::Value::key(.1.get_end_key()),
+        .1.get_id()
+    )]
+    KeyNotInRegion(Vec<u8>, metapb::Region),
+    #[error("peer {1} is not ready, region {0} max_ts {2}")]
+    DataIsNotReady(
+        u64, // region_id
+        u64, // peer_id
+        u64, // safe_ts
+    ),
```

Comment:
Better to use a struct variant here

Reply:
Actually, I think that is true for most of these variants

Reply:
Done